dialogue-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientBlockingChannel.java

```diff
@@ -0,0 +1,199 @@
+
+        if (request.body().isPresent()) {
+            Preconditions.checkArgument(
+                    endpoint.httpMethod() != HttpMethod.GET, "GET endpoints must not have a request body");
+            RequestBody body = request.body().get();
+            builder.setEntity(new RequestBodyEntity(body));
+        }
+        return new HttpClientResponse(client.execute(builder.build()));
+    }
+
+    private static final class HttpClientResponse implements Response {
+
+        private final CloseableHttpResponse response;
+        private Map<String, List<String>> headers;
+
+        HttpClientResponse(CloseableHttpResponse response) {
+            this.response = response;
+        }
+
+        @Override
+        public InputStream body() {
+            try {
+                return response.getEntity().getContent();
+            } catch (IOException e) {
+                throw new SafeRuntimeException("Failed to get response stream", e);
+            }
+        }
+
+        @Override
+        public int code() {
+            return response.getStatusLine().getStatusCode();
+        }
+
+        @Override
+        public Map<String, List<String>> headers() {
+            if (headers == null) {
+                Map<String, List<String>> tmpHeaders = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+                for (Header header : response.getAllHeaders()) {
+                    String value = header.getValue();
+                    if (value != null) {
+                        tmpHeaders
+                                .computeIfAbsent(header.getName(), _name -> new ArrayList<>(1))
+                                .add(header.getValue());
+                    }
+                }
+                headers = tmpHeaders;
+            }
+            return headers;
+        }
+
+        @Override
+        public Optional<String> getFirstHeader(String header) {
+            return Optional.ofNullable(response.getFirstHeader(header)).map(Header::getValue);
+        }
+
+        @Override
+        public void close() {
+            try {
+                response.close();
+            } catch (IOException | RuntimeException e) {
+                log.warn("Failed to close response", e);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "HttpClientResponse{response=" + response + '}';
+        }
+    }
+
+    private static final class RequestBodyEntity implements HttpEntity {
+
+        private final RequestBody requestBody;
+        private final Header contentType;
+
+        RequestBodyEntity(RequestBody requestBody) {
+            this.requestBody = requestBody;
+            this.contentType = new BasicHeader(Headers.CONTENT_TYPE, requestBody.contentType());
+        }
+
+        @Override
+        public boolean isRepeatable() {
+            return false;
+        }
+
+        @Override
+        public boolean isChunked() {
+            return true;
```

Comment:
Out of curiosity, are there ever worlds where we explicitly wouldn't want chunked? Do people just do it when you know the request size and want to save the bytes in between each chunk?

Reply:
That's correct. There's an optimization where we could internally buffer data, and if we hit the end of content before something like 8k or 16k bytes we could send a content-length request.
We already have relatively large buffers configured at the OS level so we're fine writing to the socket, we just can't walk back to the headers and add a content-length.